1. Create a 6x6 NumPy array with random integers (0-255) to simulate a grayscale image, then apply a 2x2 max pooling operation with a stride of 2 and print the resulting reduced array.


In [ ]:
import numpy as np

np.random.seed(42)

image = np.random.randint(0, 256, size=(6, 6))

print("Original 6x6 Image:")
print(image)

def max_pooling(image, pool_size=2, stride=2):
    h, w = image.shape

    out_h = (h - pool_size) // stride + 1
    out_w = (w - pool_size) // stride + 1

    pooled = np.zeros((out_h, out_w), dtype=int)

    for i in range(out_h):
        for j in range(out_w):
            region = image[
                i * stride:i * stride + pool_size,
                j * stride:j * stride + pool_size
            ]
            pooled[i, j] = np.max(region)

    return pooled

pooled_image = max_pooling(image)

print("\nPooled 3x3 Image:")
print(pooled_image)

Original 6x6 Image:
[[102 179  92  14 106  71]
 [188  20 102 121 210 214]
 [ 74 202  87 116  99 103]
 [151 130 149  52   1  87]
 [235 157  37 129 191 187]
 [ 20 160 203  57  21 252]]

Pooled 3x3 Image:
[[188 121 214]
 [202 149 103]
 [235 203 252]]


2. Write a Python function average_pooling_2x2(image) that takes a 2D NumPy array and returns a new array after applying 2x2 average pooling with stride 2.

In [ ]:
import numpy as np

def average_pooling_2x2(image):
    rows, cols = image.shape

    out_rows = rows // 2
    out_cols = cols // 2

    pooled = np.zeros((out_rows, out_cols))

    for i in range(out_rows):
        for j in range(out_cols):
            region = image[i*2:i*2+2, j*2:j*2+2]
            pooled[i, j] = np.mean(region)

    return pooled

image = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9,10,11,12],
    [13,14,15,16]
])

result = average_pooling_2x2(image)

print("Input Image:")
print(image)

print("\nAverage Pooled Output:")
print(result)

Input Image:
[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]

Average Pooled Output:
[[ 3.5  5.5]
 [11.5 13.5]]


3. Given a 4x4 feature map array, manually perform both max pooling and average pooling with a 2x2 window and stride 2, and compare the results side by side.<br><br><em><strong>Hint:</strong> Show both resulting arrays and explain in one line which pooling method keeps more details from the original image.</em>

et's use the following 4×4 feature map:

[[ 1,  3,  2,  4],
 [ 5,  6,  7,  8],
 [ 9, 10, 11, 12],
 [13, 14, 15, 16]]
Step 1: Max Pooling (2×2, Stride 2)

Divide the feature map into four 2×2 regions:

[[1, 3],     [[2, 4],
 [5, 6]]      [7, 8]]

Max = 6      Max = 8


[[ 9,10],    [[11,12],
 [13,14]]     [15,16]]

Max = 14     Max = 16

Max Pooled Output:

[[ 6,  8],
 [14, 16]]
Step 2: Average Pooling (2×2, Stride 2)

Compute the average of each 2×2 region:

(1+3+5+6)/4   = 3.75
(2+4+7+8)/4   = 5.25

(9+10+13+14)/4  = 11.5
(11+12+15+16)/4 = 13.5

Average Pooled Output:

[[ 3.75,  5.25],
 [11.50, 13.50]]
Results Side by Side
Max Pooling	Average Pooling
[[ 6, 8],	[[ 3.75, 5.25],
[14,16]]	[11.5,13.5]]

4. Combine a single 3x3 convolution layer (use any simple filter, e.g., edge detection) followed by a 2x2 max pooling layer on a 5x5 input array. Show the output after each step.


In [3]:
import numpy as np

input_array = np.array([
    [1, 2, 3, 0, 1],
    [4, 5, 6, 1, 2],
    [7, 8, 9, 2, 3],
    [1, 2, 3, 4, 5],
    [6, 7, 8, 9, 0]
])

kernel = np.array([
    [-1, -1, -1],
    [-1,  8, -1],
    [-1, -1, -1]
])

conv_output = np.zeros((3, 3))

for i in range(3):
    for j in range(3):
        region = input_array[i:i+3, j:j+3]
        conv_output[i, j] = np.sum(region * kernel)

pool_output = np.zeros((1, 1))

for i in range(0, conv_output.shape[0] - 1, 2):
    for j in range(0, conv_output.shape[1] - 1, 2):
        pool_output[i // 2, j // 2] = np.max(conv_output[i:i+2, j:j+2])

print("Input Array:")
print(input_array)

print("\nAfter Convolution:")
print(conv_output)

print("\nAfter 2x2 Max Pooling:")
print(pool_output)

Input Array:
[[1 2 3 0 1]
 [4 5 6 1 2]
 [7 8 9 2 3]
 [1 2 3 4 5]
 [6 7 8 9 0]]

After Convolution:
[[  0.  18. -18.]
 [ 27.  41. -17.]
 [-33. -25.  -7.]]

After 2x2 Max Pooling:
[[41.]]


5. Explain in your own words, using an example from Instagram or Zomato, how pooling layers help reduce computation and improve feature extraction in deep learning models.

Pooling layers help a deep learning model reduce the size of feature maps, which means fewer calculations, less memory usage, and faster training. They also help the model focus on the most important features while ignoring small variations or noise.

Instagram Example:
When Instagram detects objects in a photo (such as a dog or a face), the early convolution layers find features like edges, textures, and shapes. A max pooling layer keeps only the strongest feature values from small regions. For example, if several nearby pixels indicate the edge of a face, pooling keeps the strongest edge response and discards the rest. This reduces the amount of data the network must process while still preserving the important information needed for recognition.

Zomato Example:
Suppose Zomato uses an image recognition model to identify dishes from uploaded food photos. The convolution layers detect features such as plate boundaries, toppings, and food textures. Pooling reduces the image representation by keeping the most significant features, making the model faster and less sensitive to small changes in image position or lighting. As a result, the model can still recognize a pizza even if it appears in a slightly different location within the image.

In summary: Pooling layers reduce computation by shrinking feature maps and improve feature extraction by retaining the most important information while removing redundant details.